Go To Button

luc-ass opened the pull request:
### Added
- configuration for optional button with: name, x-coordinate, y-coordinate
- logic to send robot vacuum to location
- explanation to make finding the right coordinates somewhat easier
### Known quirks
- You need to start/stop the vacuum to send it back to its dock. My way is actually pressing the dock button on the vacuum, as I send it to the bin anyway...
- The default name of the button is highly opinionated - after all that's why I built this 😅 

Please let me know if I can improve it somehow.

luc-ass commented:
This resolves #306, #256 and #130.

luc-ass commented on index.js:
```diff
@@ -14,6 +14,7 @@ const PLUGIN_NAME = "homebridge-xiaomi-roborock-vacuum";
 const ACCESSORY_NAME = "XiaomiRoborockVacuum";
 
 const MODELS = require("./models");
+const { runInThisContext } = require("vm");
```
no idea where this came from

luc-ass replied:
resolved

luc-ass commented on miio/lib/devices/vacuum.js:
```diff
@@ -332,6 +332,14 @@ module.exports = class extends Vacuum.with(
+   * Send the vacuum to a specific location.
+  */
+  sendToLocation(x,y) {
+    console.log("called sendToLocation()")
```
might be removed if not needed

luc-ass replied:
✅ resolved

luc-ass commented:
Would this be a solution for the "button always off" problem?

https://github.com/homebridge-xiaomi-roborock-vacuum/homebridge-xiaomi-roborock-vacuum/blob/master/miio/lib/devices/vacuum.js#L225-L230

luc-ass commented on index.js:
```diff
@@ -1604,6 +1604,24 @@ class XiaomiRoborockVacuum {
+      throw err;
+    }
+  }
+
```
does now show the button as `on` while the vacuum is driving towards location

luc-ass commented on miio/lib/devices/vacuum.js:
```diff
@@ -77,6 +77,8 @@ module.exports = class extends Vacuum.with(
         case 15:
           return "docking";
+        case 16:
+          return "going-to-location"
```
16: pin&go

luc-ass commented on index.js:
```diff
@@ -1576,6 +1591,37 @@ class XiaomiRoborockVacuum {
+    }
+  }
+
+  async getGoToState() {
```
Added feedback for the button. `on` while moving to target.

luc-ass commented on config.schema.json:
```diff
@@ -69,6 +69,39 @@
+        "default": "time to empty",
+        "required": true,
+        "condition": {
+          "functionBody": "return model.goTo === true"
```
only show advanced configuration if `go to` is enabled

luc-ass replied:
I will happily add this to other parts of the configuration, if you wish 🎁

afharo commented on README.md:
```diff
@@ -154,6 +154,13 @@ This feature seems to be working on models that support naming the rooms in the
 Even if you have one of these models but you haven't named the Rooms in your App yet, this function will not work! Thanks @domeOo
 
+## Go to Location
+This plugin allowes you to send your vacuum to a specific point on the map. As there is no easy way of figuring out the coordinates of your desired location, here are some steps to help you find the right spot.
```
typo 😇 
```suggestion
This plugin allows you to send your vacuum to a specific point on the map. As there is no easy way of figuring out the coordinates of your desired location, here are some steps to help you find the right spot.
```

afharo commented on README.md:
```diff
@@ -154,6 +154,13 @@ This feature seems to be working on models that support naming the rooms in the
+## Go to Location
+This plugin allowes you to send your vacuum to a specific point on the map. As there is no easy way of figuring out the coordinates of your desired location, here are some steps to help you find the right spot.
+1. The base location (typically the position of the dock when the map was created) is at `[25500,25500]` where these values are in the format `[x-axis/horizontal, y-axis/vertical]`.
+2. Increasing the first value results in going to the right, increasing the second value results in going to the top of the map (as displayed inside the app). The value seems to be in Millimeters.
```
Reading points 1 and 3 makes me wonder if it should be
```suggestion
2. Increasing the first value results in going to the top of the map, increasing the second value results in going to the right of the map (as displayed inside the app). The value seems to be in Millimeters.
```

Did I miss something?

afharo commented on index.js:
```diff
@@ -106,6 +106,10 @@ class XiaomiRoborockVacuum {
     this.config.findMe = config.findMe || false;
     this.config.findMeWord = config.findMeWord || "where are you";
+    this.config.goTo = config.goTo || false;
+    this.config.goToWord = config.goToWord || "time to empty";
```
I love your use case, but probably `Go to coordinates` is more generic to everyone :)

luc-ass replied:
Good idea. I was out of ideas yesterday evening 😉

afharo reviewed (commented):
Wow! This is a very requested feature! Thank you @luc-ass, for taking the time to implement it 🧡 

I added a couple of minor comments.

I also have a small suggestion if you think you can make it:
Is there a possibility to create multiple `Go To` buttons? Do you think it's worth implementing this feature that way?

luc-ass commented:
> Wow! This is a very requested feature! Thank you @luc-ass, for taking the time to implement it 🧡

Sure thing! Can't wait to use the automation myself!
> I added a couple of minor comments.

Thank you for your suggestions. I implemented them.
> I also have a small suggestion if you think you can make it: Is there a possibility to create multiple `Go To` buttons? Do you think it's worth implementing this feature that way?

As of now I have no idea how to do that in an efficient way. But I will happily keep working on it when I find some more time. 😉

afharo commented:
We can merge this for now with one button and work on an improvement in the future. How does that sound?

afharo reviewed (approved):
LGTM! Thank you @luc-ass!

I'll go ahead, merge it and publish it.

We can have a follow up in the future to add multiple buttons.
